src/mpi/coll/transports/generic/transport.c

```diff
@@ -0,0 +1,1243 @@
+        else
+            *is_commutative = 1;
+    }
+}
+
+/* Transport function to tell if the buffer is in_place - used by some
+ * collective operations */
+int MPIR_COLL_GENERIC_isinplace(const void *buf)
+{
+    if (buf == MPI_IN_PLACE)
+        return 1;
+
+    return 0;
+}
+
+/* Transport function to get datatype information */
+void MPIR_COLL_GENERIC_dtinfo(MPIR_COLL_GENERIC_dt_t dt,
+                                   int *iscontig,
+                                   size_t * size, size_t * out_extent, size_t * lower_bound)
+{
+    MPI_Aint true_lb, extent, true_extent, type_size;
+
+    MPIR_Datatype_get_size_macro(dt, type_size);
+    MPIR_Type_get_true_extent_impl(dt, &true_lb, &true_extent);
+    MPIR_Datatype_get_extent_macro(dt, extent);
+    MPIR_Datatype_is_contig(dt, iscontig);
+
+    *size = type_size;
+    *out_extent = MPL_MAX(extent, true_extent);
+    *lower_bound = true_lb;
+}
+
+/* Transport function to add reference to a datatype */
+void MPIR_COLL_GENERIC_addref_dt(MPIR_COLL_GENERIC_dt_t dt, int up)
+{
+    MPIR_Datatype *dt_ptr;
+
+    if (HANDLE_GET_KIND(dt) != HANDLE_KIND_BUILTIN) {
+        MPIR_Datatype_get_ptr(dt, dt_ptr);
+
+        if (up)
+            MPIR_Datatype_add_ref(dt_ptr);
+        else
+            MPIR_Datatype_release(dt_ptr);
+    }
+}
+
+/* Internal transport function to allocate memory associated with a graph vertex */
+MPL_STATIC_INLINE_PREFIX void MPIR_COLL_GENERIC_allocate_vtx(MPIR_COLL_GENERIC_vtx_t * vtx, int invtcs_array_size,
+                                         int outvtcs_array_size)
+{
+    /* allocate memory for storing incoming and outgoing vertices */
+    vtx->invtcs.array = MPIR_COLL_GENERIC_allocate_mem(sizeof(int) * invtcs_array_size);
+    vtx->invtcs.size = invtcs_array_size;
+    vtx->outvtcs.array = MPIR_COLL_GENERIC_allocate_mem(sizeof(int) * outvtcs_array_size);
+    vtx->outvtcs.size = outvtcs_array_size;
+}
+
+/* Internal transport function to initialize a graph vertex */
+MPL_STATIC_INLINE_PREFIX void MPIR_COLL_GENERIC_init_vtx(MPIR_COLL_GENERIC_sched_t * sched, MPIR_COLL_GENERIC_vtx_t * vtx, int id)
+{
+    /* allocate memory for storing incoming and outgoing edges */
+    MPIR_COLL_GENERIC_allocate_vtx(vtx, sched->max_edges_per_vtx, sched->max_edges_per_vtx);
+    vtx->invtcs.used = 0;
+    vtx->outvtcs.used = 0;
+
+    vtx->state = MPIR_COLL_GENERIC_STATE_INIT;
+    vtx->id = id;
+    vtx->num_unfinished_dependencies = 0;
+}
+
+/* Internal transport function to free the memory associated with a vertex */
+MPL_STATIC_INLINE_PREFIX void MPIR_COLL_GENERIC_free_vtx(MPIR_COLL_GENERIC_vtx_t * vtx)
+{
+    MPIR_COLL_GENERIC_free_mem(vtx->invtcs.array);
+    MPIR_COLL_GENERIC_free_mem(vtx->outvtcs.array);
+}
+
+/* Internal transport function to add list of integers to an integer array (MPIR_COLL_GENERIC_int_array)*/
+MPL_STATIC_INLINE_PREFIX void MPIR_COLL_GENERIC_add_elems_int_array(MPIR_COLL_GENERIC_int_array * in, int n_elems, int *elems)
```

Comment:
Remove namespace

Reply:
If we convert to using a `utarray`, I don't think this function is necessary anymore.

Reply:
Yes, I will look into utarray.